API Generator Instructions Missing / Broken

Hi team!
I'm trying to run the API generator for a new API with a private discovery document. However, I'm running into a few issues.

- I can't find documentation on how to run the generator. I think it would be really useful to write this down somewhere in a README. I don't use this package enough to remember the commands.
- I'm not able to generate after a clone, `npm i`, and `npm run generate`. I get errors like this:
  -  client/samples/youtube/upload.js
    -  eslint errors: `23:26  error  "googleapis" is not found  node/no-missing-require`
    - `✖ 40 problems (40 errors, 0 warnings)`
    - I had to rmrf `samples/` for the module to build.
- There are weird fetch and access errors while generating APIs. Those should probably shouldn't be there.

`npm run generate` used to work.

Like `npm run generate https://slides.googleapis.com/$discovery/rest?version=v1`.

Now running the command fails:

```
(node:82838) UnhandledPromiseRejectionWarning: Error: <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 404 (Not Found)!!1</title>
  <style>
    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;margin-left:-5px}@media only screen and (min-resolution:192dpi){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;-moz-border-image:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0}}@media only screen and (-webkit-min-device-pixel-ratio:2){#logo{background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;-webkit-background-size:100% 100%}}#logo{display:inline-block;height:54px;width:150px}
  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>404.</b> <ins>That’s an error.</ins>
  <p>The requested URL <code>/rest?version=v1</code> was not found on this server.  <ins>That’s all we know.</ins>

    at Gaxios._request (/Users/timmerman/.../node_modules/gaxios/build/src/gaxios.js:85:23)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async run (/Users/timmerman/.../node_modules/p-queue/dist/index.js:255:29)
(node:82838) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:82838) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

> googleapis@48.0.0 postgenerate /...
> npm run fix


> googleapis@48.0.0 fix /...
> eslint --fix '**/*.js' && gts fix


Oops! Something went wrong! :(

ESLint: 6.8.0.

You are linting "**/*.js", but all of the files matching the glob pattern "**/*.js" are ignored.

If you don't want to lint these files, remove the pattern "**/*.js" from the list of arguments passed to ESLint.

If you do want to lint these files, try the following solutions:

* Check your .eslintignore file, or the eslintIgnore property in package.json, to ensure that the files are not configured to be ignored.
* Explicitly list the files from this glob that you'd like to lint on the command-line, rather than providing a glob as an argument.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! googleapis@48.0.0 fix: `eslint --fix '**/*.js' && gts fix`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the googleapis@48.0.0 fix script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/.../.npm/_logs/2020-04-10T01_39_35_319Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! googleapis@48.0.0 postgenerate: `npm run fix`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the googleapis@48.0.0 postgenerate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/.../.npm/_logs/2020-04-10T01_39_35_362Z-debug.log
```

Am I missing something? I'm guessing it's just broken and there are no tests. Is there a guide for generating this client for APIs that don't get picked up by the discovery list?

--- a/generator.md
+++ b/generator.md
@@ -1,0 +1,50 @@
+# Using the Generator
+The clients in this repository are all automatically generated.  They are generated using the [Google Discovery Service](https://developers.google.com/discovery).  Most users of this library will not need to directly use the generator.  This documentation is intended for users that need to maintain the repository, or build custom clients.
+
+## Running the generator locally
+To run the generator locally:
+1. Clone the repository, and cd into the directory
+2. Run `npm install`
+3. Run `npm run generate`
+
+This command will download all discovery files available via the [discovery index](https://www.googleapis.com/discovery/v1/apis/), and run the code generator against those files.
+
+### Generatoring from the cache
+There are some situations where you don't want to re-download the discovery files before running the generator.  This is particularly useful in situations where you're working on the generator itself, and want to isolate the change.  To prevent the generator from re-downloading the discovery files, run:
+```sh
+npm run generate -- --use-cache
+```
+
+### Downloading discovery files
+As part of the `npm run generate` command, discovery docs are downloaded.  To only download discovery file updates, and not re-run the generator, run:
+```sh
+npm run download
+```
+
+## Submitting generator PRs
+This repository uses [synthtool](https://github.com/googleapis/synthtool/) to re-generate the API on a nightly basis.  The command `npm run submit-prs` will perform a variety of steps:
+1. Download all discovery files (`npm run download`)
+2. Run the generator (`npm run generate`)
+3. Iterate over each directory in `src/apis`, and create a commit with only those changes
+4. Submit a single PR with multiple commits, including a changelog
+
+## Generating individual APIs
+You can generate a single API based on a discovery URL. Replace the url and API name below to match the API you'd like to generate:
+```
+npm run build-tools
+node build/src/generator/generator.js 'https://apigee.googleapis.com/$discovery/rest?version=v1'
+```
+
+Assuming that command completes successfully, you can access the output and test it out by navigating to the API subdirectory:
+```
+cd src/apis/apigee
+npm install
+```
+
+If you want to distribute this package, which is common for APIs not listed in the discovery index, you can pack it to generate a tarball:
+```
+npm pack
+```
+
+## Questions
+Feel free to submit an issue!

--- a/src/generator/generator.ts
+++ b/src/generator/generator.ts
@@ -17,11 +17,11 @@ import {Schema, Schemas} from 'googleapis-common';
 import * as mkdirp from 'mkdirp';
 import * as nunjucks from 'nunjucks';
 import * as path from 'path';
-import {URL} from 'url';
 import * as util from 'util';
 import Q from 'p-queue';
 import * as prettier from 'prettier';
 import * as minimist from 'yargs-parser';
+import {request} from 'gaxios';
 import {DISCOVERY_URL} from './download';
 import {downloadDiscoveryDocs, ChangeSet} from './download';
 import * as filters from './filters';
@@ -218,21 +218,15 @@ export class Generator {
    * @param apiDiscoveryUri URL or filename of discovery doc for API
    */
   async generateAPI(apiDiscoveryUrl: string) {
-    let parts: URL;
-    try {
-      parts = new URL(apiDiscoveryUrl);
-    } catch (e) {
-      // 🤷‍♂️
-    }
-    if (apiDiscoveryUrl && !parts!) {
+    const isUrl = apiDiscoveryUrl.startsWith('https://');
+    if (!isUrl) {
       this.log(`Reading from file ${path.relative('.', apiDiscoveryUrl)}`);
       const file = await readFile(apiDiscoveryUrl, 'utf-8');
       await this.generate(apiDiscoveryUrl, JSON.parse(file));
     } else {
-      this.logResult(apiDiscoveryUrl, apiDiscoveryUrl);
-      const file = await readFile(apiDiscoveryUrl, 'utf8');
-      const data = JSON.parse(file) as Schema;
-      await this.generate(apiDiscoveryUrl, data);
+      this.log(`Reading from url ${apiDiscoveryUrl}`);
+      const res = await request<Schema>({url: apiDiscoveryUrl});
+      await this.generate(apiDiscoveryUrl, res.data);
     }
   }
 
@@ -281,6 +275,7 @@ async function main() {
   const gen = new Generator({debug: true, includePrivate: false});
   if (!discoveryUrl && argv._.length > 0) {
     argv._.forEach(async url => {
+      console.log(`Generating API for ${url}`);
       await gen.generateAPI(url);
       console.log('Generated API for ' + url);
     });

